pretrain/tinyllama.py

```diff
@@ -0,0 +1,341 @@
+
+
+def create_dataloaders(batch_size: int, block_size: int) -> Tuple[DataLoader, DataLoader]:
+    from lightning.data import StreamingDataset
+    from lightning.data.streaming.item_loader import TokensLoader
+
+    # Increase by one because we need the next word as well
+    effective_block_size = block_size + 1
+
+    train_datasets = [
+        StreamingDataset(
+            input_dir="data/slimpajama/train",
+            item_loader=TokensLoader(block_size=effective_block_size), 
+            shuffle=True,
+            drop_last=True,
+        ),
+        StreamingDataset(
+            input_dir="data/starcoder",
+            item_loader=TokensLoader(block_size=effective_block_size), 
+            shuffle=True,
+            drop_last=True,
+        ),
+    ]
+
+    # Mix SlimPajama data and Starcoder data with these proportions:
+    weights = (0.693584, 0.306416)
+    combined_dataset = CombinedDataset(datasets=train_datasets, seed=42, weights=weights)
+    train_dataloader = DataLoader(combined_dataset, batch_size=batch_size, pin_memory=True, num_workers=8)
+
+    val_dataset = StreamingDataset(
+        input_dir="data/slimpajama/val",
+        item_loader=TokensLoader(block_size=effective_block_size), 
+        shuffle=False,
+        # Consider setting to False, but we would lose some samples due to truncation when world size > 1
+        drop_last=True,
+    )
+    val_dataloader = DataLoader(val_dataset, batch_size=batch_size, pin_memory=True, num_workers=8)
+    return train_dataloader, val_dataloader
+
+
+# learning rate decay scheduler (cosine with warmup)
+def get_lr(it):
+    # 1) linear warmup for warmup_iters steps
+    if it < warmup_iters:
+        return learning_rate * it / warmup_iters
+    # 2) if it > lr_decay_iters, return min learning rate
+    if it > lr_decay_iters:
+        return min_lr
+    # 3) in between, use cosine decay down to min learning rate
+    decay_ratio = (it - warmup_iters) / (lr_decay_iters - warmup_iters)
+    assert 0 <= decay_ratio <= 1
+    coeff = 0.5 * (1.0 + math.cos(math.pi * decay_ratio))  # coeff ranges 0..1
+    return min_lr + coeff * (learning_rate - min_lr)
+
+
+def init_weights(module: nn.Module, n_layer: int):
+    # Follows GPT-NeoX: https://arxiv.org/abs/2204.06745
+    if isinstance(module, nn.Embedding):
+        nn.init.normal_(module.weight, mean=0.0, std=math.sqrt(2.0 / 5 / module.weight.shape[1]))
+    elif isinstance(module, nn.Linear):
+        nn.init.normal_(module.weight, mean=0.0, std=math.sqrt(2.0 / 5 / module.weight.shape[1]))
+        if module.bias is not None:
+            nn.init.zeros_(module.bias)
+    for name, param in module.named_parameters():
+        if name == "proj.weight" and isinstance(module, LLaMAMLP):
+            nn.init.normal_(param, mean=0.0, std=(1 / math.sqrt(param.shape[-1]) / n_layer))
+
+
+if __name__ == "__main__":
+    torch.set_float32_matmul_precision("high")
+
+    from jsonargparse import CLI
+
+    CLI(setup)
```

Comment:
Since the script uses torch.compile by default and it requires 2.2, I would check that it's available here